lib/config/flat-config-schema.js

```diff
@@ -0,0 +1,362 @@
+ */
+function assertIsObject(value) {
+    if (!value || typeof value !== "object") {
+        throw new TypeError("Expected an object.");
+    }
+}
+
+/**
+ * Validates that a value is an object or a string.
+ * @param {any} value The value to check.
+ * @returns {void}
+ * @throws {TypeError} If the value isn't an object or a string.
+ */
+function assertIsObjectOrString(value) {
+    if ((!value || typeof value !== "object") && typeof value !== "string") {
+        throw new TypeError("Expected an object or string.");
+    }
+}
+
+//-----------------------------------------------------------------------------
+// Low-Level Schemas
+//-----------------------------------------------------------------------------
+
+/** @type {ObjectPropertySchema} */
+const objectAssignSchema = {
+    merge: "assign",
+    validate: "object"
+};
+
+/** @type {ObjectPropertySchema} */
+const numberSchema = {
+    merge: "replace",
+    validate: "number"
+};
+
+/** @type {ObjectPropertySchema} */
+const booleanSchema = {
+    merge: "replace",
+    validate: "boolean"
+};
+
+//-----------------------------------------------------------------------------
+// High-Level Schemas
+//-----------------------------------------------------------------------------
+
+/** @type {ObjectPropertySchema} */
+const globalsSchema = {
+    merge: "assign",
+    validate(value) {
+
+        assertIsObject(value);
+
+        for (const key of Object.keys(value)) {
+
+            // avoid hairy edge case
+            if (key === "__proto__") {
+                continue;
+            }
+
+            if (key !== key.trim()) {
+                throw new TypeError(`Global "${key}" has leading or trailing whitespace.`);
+            }
+
+            if (!globalVariablesValues.has(value[key])) {
+                throw new TypeError("Expected \"readonly\", \"writeable\", or \"off\".");
+            }
+        }
+    }
+};
+
+/** @type {ObjectPropertySchema} */
+const parserSchema = {
+    merge: "replace",
+    validate(value) {
+        assertIsObjectOrString(value);
+
+        if (typeof value === "object" && typeof value.parse !== "function" && typeof value.parseForESLint !== "function") {
+            throw new TypeError("Expected object to have a parse() or parseForESLint() method.");
+        }
+
+        if (typeof value === "string") {
+            assertIsPluginMemberName(value);
+        }
+    }
+};
+
+/** @type {ObjectPropertySchema} */
+const pluginsSchema = {
+    merge(first = {}, second = {}) {
+        const keys = new Set([...Object.keys(first), ...Object.keys(second)]);
+        const result = {};
+
+        // manually validate that plugins are not redefined
+        for (const key of keys) {
+
+            // avoid hairy edge case
+            if (key === "__proto__") {
+                continue;
+            }
+
+            if (key in first && key in second && first[key] !== second[key]) {
+                throw new TypeError(`Cannot redefine plugin "${key}".`);
```

Comment:
There's not currently a test exercising this check.

Reply:
Tested on line 260.